feat: search docs (#2297)
ory/docs: docs/kratos/manage-identities/search/identity-search-api.mdx

[Get in touch](https://www.ory.sh/contact/) for more information.

:::

Advanced identity search allows you to perform full-text search on identity traits, metadata (public+admin), combined with
arbitrary filters and faceting.

## Getting started

We recommend using the [Typesense SDKs](https://typesense.org/docs/29.0/api/api-clients.html) for your preferred language to
interact with the advanced identity search API. Pass your [Ory Network API key](/concepts/personal-access-token) where you would
normally pass the Typesense API key, use the following base URL:

```shell
https://{your-ory-slug}.projects.oryapis.com/admin/preview/search/v0beta1
```

and set the collection name to `identities-{your-project-id}`.

See the following examples for how to configure and use the Typesense SDKs to search for identities.

```mdx-code-block
<Tabs>
  <TabItem value="shell" label="cURL" default>
    <CodeBlock language="shell">{basic}</CodeBlock>
  </TabItem>
  <TabItem value="go" label="Go">
    <CodeBlock language="go">{goExample}</CodeBlock>
  </TabItem>
  <TabItem value="java" label="Java">
    <CodeBlock language="clike">{javaExample}</CodeBlock>
  </TabItem>
</Tabs>
```

## Data model

> The collection `identities-{your-project-id}` contains one document per identity. Each document contains the following fields:

Reviewer: The collection `identities-{your-project-id}` -> The `identities-{your-project-id}` collection

> The collection `identities-{your-project-id}` contains one document per identity. Each document contains the following fields:

Reviewer: collection, search index, or collection/search-index?

```mdx-code-block
<CodeBlock language="json5">{model}</CodeBlock>
```

:::tip

> To avoid timing out search requests, specify the exact fields you want to search in using the `query_by` parameter. For example,

Reviewer: To avoid timing out search requests - > To avoid search requests timing out
search in using -> search by using

to search for identities by email, use `query_by=traits.email`.

:::

:::note

Because of technical limitations, non-string data in `metadata_admin` and `metadata_public` will be indexed as strings and
returned from the Search API as strings.

To retrieve the original JSON data, fetch the full identity using the
[Get Identity API](/reference/api#tag/identity/operation/getIdentity).

:::

## Advanced search example

Please refer to the [Typesense documentation](https://typesense.org/docs/29.0/api/search.html) for details on how to construct

> arbitrary search queries. An example:

Reviewer: arbitrary -> custom

```mdx-code-block

<Tabs>
  <TabItem value="curl" label="cURL" default>
    <CodeBlock language="shell">{advanced}</CodeBlock>
  </TabItem>

  <TabItem value="result" label="sample result">
    <CodeBlock language="json5">{advancedResult}</CodeBlock>
  </TabItem>
</Tabs>
```

:::tip

Most fields support infix-searching as well, so queries like `query_by=@example.com&infix=always&query_by=traits.email` are
possible.

:::

## Consistency and availability

> The search index powering advanced identity search is eventually consistent with the main identity store. This means that there

Reviewer: remove "The search index powering advanced identity search is eventually consistent with the main identity store. This means that there"

> The search index powering advanced identity search is eventually consistent with the main identity store. This means that there

Reviewer: When you create, update, or delete an identity, there is a slight delay in the change being reflected in the advanced identity search results. Any changes should take less that one second for the search index to be consistent with the main identity store, during normal operations.

> The search index powering advanced identity search is eventually consistent with the main identity store. This means that there

Reviewer: According to a Claude, Typesense's 'search index' term is the same as 'collection'. We've used both terms in this document. We should pick one or if we think both are useful to use combine them. collection/search-index

> will be a delay between creating, updating, or deleting an identity and the changes being reflected in search results. We try to

Reviewer: Remove: "will be a delay between creating, updating, or deleting an identity and the changes being reflected in search results. We try to"

> aim for a delay of less than one second during normal operation, but cannot provide any guarantees. The search index will have to

Reviewer: remove: "aim for a delay of less than one second during normal operation, but cannot provide any guarantees."

> will be a delay between creating, updating, or deleting an identity and the changes being reflected in search results. We try to
> aim for a delay of less than one second during normal operation, but cannot provide any guarantees. The search index will have to
> be rebuilt periodically, during which time we cannot service search requests.

Reviewer 2: ```suggestion
will be a delay between creating, updating, or deleting an identity and the changes being reflected in search results. Ory aims for a delay of less than one second during normal operation, but there are no guarantees. The search index will be rebuilt periodically, during which time search requests are paused.
```

there is no try ;) 
also no we

> be rebuilt periodically, during which time we cannot service search requests.

Reviewer: "The search index will have to be rebuilt periodically,..." Who is performing the rebuild of the search index? Ory or the developer? Clarify who.

Change: "during which time we cannot service search requests" -> "during which time you cannot perform advanced search requests.

> We recommend you use the search API for non-time-critical use cases, such as admin UIs or background jobs, where you need to full

Reviewer: If the intent of this sentence, "We recommend you use the search API for non-time-critical use cases," is to suggest an limitation on when to use this search API, then I suggest we add 'only'.

"We recommend you use the search API only for non-time-critical use cases,"

> We recommend you use the search API for non-time-critical use cases, such as admin UIs or background jobs, where you need to full
> capability of full-text search, filtering, faceting, and sorting.

Reviewer 2: ```suggestion
It's recommended you use the search API for non-time-critical use cases, such as admin UIs or background jobs, where you need to full
capability of full-text search, filtering, faceting, and sorting.
```

> Before displaying search results to end-users or otherwise using the search results for business logic, we recommend you always
> [fetch the full identity](/reference/api#tag/identity/operation/getIdentity) from the main identity store using the identity ID
> returned in search results to ensure you have the most up-to-date information.

Reviewer 2: ```suggestion
Before displaying search results to end-users or otherwise using the search results for business logic, it's recommended you always
[fetch the full identity](/reference/api#tag/identity/operation/getIdentity) from the main identity store using the identity ID
returned in search results to ensure you have the most up-to-date information.
```
